### PR TITLE
Image hooks test #3

### PR DESCRIPTION
**Pull request by imjalpreet**
This is the test for Image Hooks.

### CODE REVIEW

**Comment by imjalpreet** [comment]
The test is still in progress.

I have pushed the code that I have written till now.
@oparoz Please tell me if it is correct.

**Comment by imjalpreet** [comment]
@oparoz I have a few questions.. 

First, how can I mock the insert method of Mapper class? Is this correct:

```
$this->mapper->expects($this->any())
                       ->method('insert')
                       ->with($imageDimension)
                       ->willReturn($modifiedImageDimension);

```

Here `$modifiedImageDimension`  is different from `$imageDimension` as insert method assigns the database row id to `$imageDimension`, i.e., `$imageDimension->setId(lastInsertId)`. Is my approach correct?

Also, how should I test `post_create`, as it does not return anything? or should I change `post_create` so that it returns the database row id?

**Comment by oparoz** [comment]
> Also, how should I test post_create, as it does not return anything

Instead of `$this->any()`, use `$this->once()`
It will fail if the method is never called or if it's called too many times.
You just want to make sure the method is called, you're not testing that it's storing data.

**Comment by oparoz** [comment]
You also need to cleanup the root folder and remove the Makefile and the phpunit files, because that's not the testing suite we're using.

**Comment by imjalpreet** [comment]
@oparoz I have changed any to once. Also, what all do I have remove from the root folder in the process of clean-up?

One more thing, what is your view on my approach to mock the insert method of Mapper class that I had mentioned in one of my previous comment?

**Comment by oparoz** [comment]
> Also, what all do I have remove from the root folder in the process of clean-up?
- Makefile
- phpunit.integration.xml
- phpunit.xml

All these files were generated by ocdev

**Comment by imjalpreet** [comment]
@oparoz What should be my next step?

**Comment by oparoz** [comment]
> What should be my next step?

The first step would be to configure Codeception so that it runs your unit tests

```
  [Codeception\Exception\ConfigurationException]  
  Configuration file could not be found.          
  Run `bootstrap` to initialize Codeception.     
```

**Comment by imjalpreet** [comment]
@oparoz I have changed as we discussed. But I still have couldn't mock getimagesize correctly, can you give some example for native php methods, on how to mock them.

**Comment by oparoz** [comment]
> can you give some example for native php methods, on how to mock them.

As discussed, use you namespace to override the built-in method:
https://startpage.com/do/search?query=phpunit+test+global+php+methods&cat=web&pl=chrome&language=english_uk

More specifically:
https://www.schmengler-se.de/en/2011/03/php-mocking-built-in-functions-like-time-in-unit-tests/

File: tests/unit/Hooks/ImageHooksTest.php [hunk 1]
```diff
@@ -202,10 +142,61 @@ protected function mockJpgFileMethods($node, $path) {
...
+		return $this->getMockBuilder('OCA\MediaMetadata\Services\ImageDimension')
+			->disableOriginalConstructor()
+			->getMock();
+	}
+
+	/**
+	 * @param array $mockedMethods
+	 * @return object|ImageHooks|\PHPUnit_Framework_MockObject_MockObject
+	 */
+	private function mockPostCreate() {
```
Comment (oparoz) [inline tests/unit/Hooks/ImageHooksTest.php]: This doesn't mock postCreate, it creates a mocked object of ImageHooks

Comment (imjalpreet) [inline tests/unit/Hooks/ImageHooksTest.php]: Yes, Okay. Actually, you had told to change the name of getImageHooks to mockPostCreate, so I had changed at that time but after that I changed the contents of this function as we do not need to mock post create in this test.  I will change it back to getImageHooks.

**Comment by imjalpreet** [comment]
> As discussed, use you namespace to override the built-in method:
> https://startpage.com/do/search?query=phpunit+test+global+php+methods&cat=web&pl=chrome&language=english_uk
> 
> More specifically:
> https://www.schmengler-se.de/en/2011/03/php-mocking-built-in-functions-like-time-in-unit-tests/

Okay, Thanks. I forgot that you told about this.

File: tests/unit/Hooks/ImageHooksTest.php (hunk 1 above)
Comment (oparoz) [inline tests/unit/Hooks/ImageHooksTest.php]: There is a misunderstanding. Your previous method was doing 2 things, so you needed to put one part in `setUp` so that you could call the methods to test (the example I gave you from Gallery) and one part in a `postCreate` mock method so that it could be called in the test of `register()`.
What you did is to put the part which was destined for setUp in the `postCreate` mock method.
There is no need to create `getImageHooks`. It's easier to read when we have everything for the main object creation in the `setUp` method, like in the Gallery example.

Comment (imjalpreet) [inline tests/unit/Hooks/ImageHooksTest.php]: Okay, Got it. I will do the required changes.

**Comment by oparoz** [comment]
And don't forget that you still need to configure Codeception to run your tests

**Comment by imjalpreet** [comment]
@oparoz Can you review the updated test? I have overridden the getimagesize method but I am still not sure about my mockGetImageSize method, can you tell me if it's correct?

I will commit the bootstrap and codeception config file in the next commit.

**Comment by imjalpreet** [comment]
@oparoz I have added the basic configuration for codeception(based on their documentation) but I haven't tested it yet. Can you tell me should I include the `extension` and `coverage` tags to my codeception.yml or not? 

I will test it once you confirm about the test and the config file.

**Comment by imjalpreet** [comment]
Also, do I need to create the _data, _output and other similar directories in the tests directory? or will it be automatically created when I run the tests for the first time?

**Comment by oparoz** [comment]
> but I haven't tested it yet

Why? It's important that you learn how to run tests locally first.

> should I include the extension and coverage tags to my codeception.yml or not?

You don't need them for now, but as soon as this PR is completed, you should create another one which adds coverage. It will be helpful when adding more methods.

> Also, do I need to create the _data, _output and other similar directories in the tests directory? or will it be automatically created when I run the tests for the first time?
-  `_data` will be for your fixtures when running integration tests
-  `_output` is automatically created
-  `_envs` will be for acceptance tests
- `_support` will contains some of you helper classes. It seems you haven't added the files to your commit, so once you do, things should work.

File: tests/unit/Hooks/ImageHooksTest.php [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+			->disableOriginalConstructor()
+			->getMock();
+		$this->imageDimension = $this->mockImageDimension();
+
+		$app = new \OCA\MediaMetadata\AppInfo\Application();
+		$container = $app->getContainer();
+		$this->imagehooks = new ImageHooks(
+			$this->root,
+			$this->mapper,
+			$container->query('ServerContainer')->getConfig()->getSystemValue('datadirectory')
```
Comment (oparoz) [inline tests/unit/Hooks/ImageHooksTest.php:111]: Why isn't this mocked? Does the test fail without a real datadirectory?

File: tests/unit/Hooks/ImageHooksTest.php [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+
+	/**
+	 * @param array $mockedMethods
+	 * @return object|ImageHooks|\PHPUnit_Framework_MockObject_MockObject
+	 */
+	private function mockPostCreate(array $mockedMethods = []) {
+		$app = new \OCA\MediaMetadata\AppInfo\Application();
+		$container = $app->getContainer();
+
+		return $this->getMockBuilder('OCA\MediaMetadata\Hooks\ImageHooks')
```
Comment (oparoz) [inline tests/unit/Hooks/ImageHooksTest.php:235]: I think that instead of recreating the mock object, you should just add the `postCreate` method to `$this->imagehooks`. Not important for this PR since we just want tests for `testPostCreate`

File: tests/unit/Hooks/ImageHooksTest.php (hunk 2 above)
Comment (imjalpreet) [inline tests/unit/Hooks/ImageHooksTest.php:111]: Yeah, it was giving an error and then I tried this and it didn't give the error again.

**Comment by imjalpreet** [comment]
@oparoz I have been trying to run the tests but earlier it was failing and giving the errors that the `methodname` expects to be called once but it is not called. Then I wrote the test again because it was not working and it started giving an error `PHP Fatal error:  Class 'OCA\MediaMetadata\AppInfo\Application' not found in /var/www/GSoC/apps/mediametadata/tests/unit/Hooks/ImageHooksTest.php on line 67`, then I came back to the old code and I am not able to figure out why but now this code is also giving the same error. Can you find some error in the code?

Also, while I try to figure out the problem with this, can you give me my next task? Can I start with the data extraction from the EXIF data? I want to start it soon but this test is not getting over and I am not able to figure out the error.

**Comment by oparoz** [comment]
> it was failing and giving the errors that the methodname expects to be called once but it is not called

You have to be more precise. It could be that  your mock for `getimagesize` is not working properly.

> it started giving an error PHP Fatal error: Class 'OCA\MediaMetadata\AppInfo\Application' not found in /var/www/GSoC/apps/mediametadata/tests/unit/Hooks/ImageHooksTest.php on line 67, then I came back to the old code and I am not able to figure out why but now this code is also giving the same error. Can you find some error in the code?

This could be a problem related to how tests are bootstrapped, since your app is using PSR-4.
I'll take a look to see if I can see what's going on.

> Also, while I try to figure out the problem with this, can you give me my next task? Can I start with the data extraction from the EXIF data? I want to start it soon but this test is not getting over and I am not able to figure out the error.

Of course, no need to ask. If you're really stuck and need help from me, while you wait, you can always look at all the other tasks which have already been created or you can create more since you know what the goal of the project is. There are also all the admin tasks which have to be performed weekly, such as time and task tracking.

**Comment by imjalpreet** [comment]
@oparoz I am not able to figure out the mistake in the code. Can you help me out? It's giving the error: `Class 'OCA\MediaMetadata\AppInfo\Application' not found in /var/www/GSoC/apps/mediametadata/tests/unit/Hooks/ImageHooksTest.php on line 67` and I don't know why am I getting this error.

File: tests/unit/Hooks/ImageHooksTest.php [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+			->with($imageHeight);
+		$this->imageDimension->expects($this->once())
+			->method('setImageWidth')
+			->with($imageWidth);
+
+		$this->mapper->expects($this->once())
+			->method('insert')
+			->with($this->imageDimension);
+
+		$this->imagehooks->post_create($jpgFile);
```
Comment (oparoz) [inline tests/unit/Hooks/ImageHooksTest.php]: This will never work since this method doesn't exist.

**Comment by oparoz** [comment]
> I am not able to figure out the mistake in the code

First you need to fix the obvious problems so that others can use your code.
This still hasn't been fixed:

> _support will contains some of you helper classes. It seems you haven't added the files to your commit, so once you do, things should work

Please add it asap so that I can test :)

File: tests/unit/Hooks/ImageHooksTest.php (hunk 4 above)
Comment (imjalpreet) [inline tests/unit/Hooks/ImageHooksTest.php]: No, actually I haven't merged this branch with master, so in this branch the method is still post_create. I will update it once everything works.

**Comment by imjalpreet** [comment]
@oparoz I thought I had already committed the _support directory but I just checked now and they are not yet committed, sorry for that. I will commit now.

**Comment by imjalpreet** [comment]
@oparoz Actually I had forgot to push the changes. I have pushed them now.

**Comment by oparoz** [comment]
> Actually I had forgot to push the changes.

Ah, yeah, that happens sometimes :)

**Comment by oparoz** [comment]
While you're there, could you clean up the tests folder by removing the tests you didn't create? For `PageController` per example.

File: tests/unit/Hooks/ImageHooksTest.php (hunk 4 above)
Comment (oparoz) [inline tests/unit/Hooks/ImageHooksTest.php]: Ah, that explained why I had to merge then. It would be great if you could rebase the branch.

Comment (imjalpreet) [inline tests/unit/Hooks/ImageHooksTest.php]: Yeah, I will, just waiting for this test to work.

**Comment by imjalpreet** [comment]
Removed.

**Comment by oparoz** [comment]
Thanks!

File: tests/_bootstrap.php [hunk 5]
```diff
@@ -0,0 +1,2 @@
+<?php
+// This is global bootstrap for autoloading
```
Comment (oparoz) [inline tests/_bootstrap.php]: This file shouldn't be empty. This is where the boostrapping happens

**Comment by oparoz** [comment]
> I am not able to figure out the mistake in the code. Can you help me out? It's giving the error: Class 'OCA\MediaMetadata\AppInfo\Application' not found in /var/www/GSoC/apps/mediametadata/tests/unit/Hooks/ImageHooksTest.php on line 67 and I don't know why am I getting this error.

Found the problem. The app is not enabled. You need to do this:
`sudo -uwww-data ./occ app:enable mediametadata`

File: tests/unit/Hooks/ImageHooksTest.php [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+use OCP\Files\File;
+use OCP\Files\Folder;
+use PHPUnit_Framework_TestCase;
+
+/**
+ * Class ImageHooksTest
+ *
+ * @package OCA\MediaMetadata\Hooks
+ */
+class ImageHooksTest extends PHPUnit_Framework_TestCase {
```
Comment (oparoz) [inline tests/unit/Hooks/ImageHooksTest.php]: You need to extend `\Test\TestCase`

File: tests/unit/Hooks/ImageHooksTest.php (hunk 4 above)
Comment (oparoz) [inline tests/unit/Hooks/ImageHooksTest.php]: It's not a good idea to wait in this case because this branch is missing composer.json which is required to install Codeception to run tests.

File: tests/unit/Hooks/ImageHooksTest.php [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+	public function testPostCreate() {
+		$location = 'testFolder/test.jpg';
+		$fileId = 260495;
+		$jpgFile = $this->mockJpgFile($location, $fileId);
+
+		$imageWidth = 100;
+		$imageHeight = 100;
+		$dimensions = array($imageWidth, $imageHeight);
+
+		$this->mockGetImageSize($this->imagehooks, $location, $dimensions);
```
Comment (oparoz) [inline tests/unit/Hooks/ImageHooksTest.php]: You can't pass the real object as an argument to a method which expects a mock object.

File: tests/unit/Hooks/ImageHooksTest.php [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+
+		$this->logger->log('debug', 'Test Image Path: '.$location, array('app' => 'MediaMetadata'));
+		$this->logger->log('debug', 'Test Image Height: '.$imageHeight, array('app' => 'MediaMetadata'));
+		$this->logger->log('debug', 'Test Image Width: '.$imageWidth, array('app' => 'MediaMetadata'));
+
+		$this->imageDimension->expects($this->once())
```
Comment (oparoz) [inline tests/unit/Hooks/ImageHooksTest.php]: There is a problem with the imageDimension mocked methods, because these don't exist, they're created on the fly. Would be worth checking in core to see how they mock `Entity`

File: tests/unit/Hooks/ImageHooksTest.php [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+			->method('setImageId')
+			->with($jpgFile->getId());
+		$this->imageDimension->expects($this->once())
+			->method('setImageHeight')
+			->with($imageHeight);
+		$this->imageDimension->expects($this->once())
+			->method('setImageWidth')
+			->with($imageWidth);
+
+		$this->mapper->expects($this->once())
```
Comment (oparoz) [inline tests/unit/Hooks/ImageHooksTest.php]: This doesn't work. The mapper is expecting a mocked object, but you're sending a real DB object.

**Comment by imjalpreet** [comment]
@oparoz I will look at all of them and get back to you.

**Comment by imjalpreet** [comment]
@oparoz I have updated everything except the entity object mock part.
I will see core to write the mock for entity but I am getting this failure message when I run the test:
`Expectation failed for method name is equal to <string:insert> when invoked 1 time(s).
Method was expected to be called 1 times, actually called 0 times.`
Any ideas what other mistakes have been done by me?

File: tests/unit/Hooks/ImageHooksTest.php [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+			->method('insert')
+			->with($this->imageDimension);
+
+		$this->mockImageHooks->postCreate($jpgFile);
```
Comment (imjalpreet) [inline tests/unit/Hooks/ImageHooksTest.php]: @oparoz I was getting a segmentation fault when I used the real object but now, when I am using the mocked object, the test is atleast running though not passing. Any idea why it was giving segfault when using real object?

Comment (oparoz) [inline tests/unit/Hooks/ImageHooksTest.php]: I'm not sure, you should be getting a stacktrace in one of your logs. The `_output` folder usually contains useful info.

File: tests/unit/Hooks/ImageHooksTest.php [hunk 11]
```diff
@@ -112,7 +117,7 @@ public function testPostCreate() {
 			->method('insert')
 			->with($this->imageDimension);
 
-		$this->imagehooks->post_create($jpgFile);
+		$this->mockImageHooks->postCreate($jpgFile);
```
Comment (oparoz) [inline tests/unit/Hooks/ImageHooksTest.php]: Why did you remove `imagehooks`? Maybe a side-effect of the rebase?

File: tests/unit/Hooks/ImageHooksTest.php [hunk 12]
```diff
@@ -92,7 +95,9 @@ public function testPostCreate() {
 		$imageHeight = 100;
 		$dimensions = array($imageWidth, $imageHeight);
 
-		$this->mockGetImageSize($this->imagehooks, $location, $dimensions);
+		$this->mockImageHooks = $this->mockPostCreate(['postCreate',]);
```
Comment (oparoz) [inline tests/unit/Hooks/ImageHooksTest.php]: Not needed as we don't need to mock `postCreate` in this test.

File: tests/unit/Hooks/ImageHooksTest.php (hunk 11 above)
Comment (imjalpreet) [inline tests/unit/Hooks/ImageHooksTest.php]: No, actually that was intentional, as I had mentioned that when I am using imagehooks it is giving a segfault but it doesn't give a segfault when I used the mock version.

Comment (imjalpreet) [inline tests/unit/Hooks/ImageHooksTest.php]: No, actually that was intentional, as I had mentioned that when I am using imagehooks it is giving a segfault but it doesn't give a segfault when I used the mock version.

**Comment by oparoz** [comment]
> but I am getting this failure message when I run the test:
> Expectation failed for method name is equal to string:insert when invoked 1 time(s). Method was expected to be called 1 times, actually called 0 times.
> Any ideas what other mistakes have been done by me?

That's because you're never calling the method to test, you've replaced the real ImageHook object with a mock.

File: tests/unit/Hooks/ImageHooksTest.php (hunk 12 above)
Comment (imjalpreet) [inline tests/unit/Hooks/ImageHooksTest.php]: Okay, I will remove that but then how can I check whether getimagesize is called exactly once or not? I am not able to figure out that.

File: tests/unit/Hooks/ImageHooksTest.php (hunk 10 above)
Comment (imjalpreet) [inline tests/unit/Hooks/ImageHooksTest.php]: I will have a look in the _output folder.

**Comment by imjalpreet** [comment]
> That's because you're never calling the method to test, you've replaced the real ImageHook object with a mock.

Yeah, I understood. But the problem is when I am using real ImageHook object, I am getting segmentation fault and in the _output folder there is just one file names 'failed' with `tests/unit/Hooks/ImageHooksTest.php:testPostCreate` in it.

File: tests/unit/Hooks/ImageHooksTest.php (hunk 11 above)
Comment (oparoz) [inline tests/unit/Hooks/ImageHooksTest.php]: Ah, sorry, I didn't understand that. I didn't notice you were talking about `ImageHooks` I'm not getting a segfault here, so maybe a PHP update went bad or a module was recently updated. The logs should tell you what went wrong.

Comment (imjalpreet) [inline tests/unit/Hooks/ImageHooksTest.php]: Which logs are you talking about?

File: tests/unit/Hooks/ImageHooksTest.php (hunk 12 above)
Comment (oparoz) [inline tests/unit/Hooks/ImageHooksTest.php]: You'll know because you're checking that `imageDimension` methods are called once, using the image dimension as arguments.
You see what I mean?

File: tests/unit/Hooks/ImageHooksTest.php (hunk 11 above)
Comment (oparoz) [inline tests/unit/Hooks/ImageHooksTest.php]: PHP logs

File: tests/unit/Hooks/ImageHooksTest.php (hunk 12 above)
Comment (imjalpreet) [inline tests/unit/Hooks/ImageHooksTest.php]: Ohkay, I see. So, basically I can remove the mockGetImageSize method as it is not required.

File: tests/unit/Hooks/ImageHooksTest.php (hunk 11 above)
Comment (imjalpreet) [inline tests/unit/Hooks/ImageHooksTest.php]: This is the error written in syslog: `Jun 30 23:54:13 imjalpreet kernel: [ 1406.717116] php[4468]: segfault at 7fff78dc6ff8 ip 000000000079a1e6 sp 00007fff78dc7000 error 6 in php5[400000+7f4000]` but there isn't any php log file in `/var/log/`

File: tests/unit/Hooks/ImageHooksTest.php (hunk 12 above)
Comment (oparoz) [inline tests/unit/Hooks/ImageHooksTest.php]: I think so, so maybe comment it out for now.

File: tests/unit/Hooks/ImageHooksTest.php [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+ * later. See the COPYING file.
+ *
+ * @author Jalpreet Singh Nanda (:imjalpreet) <jalpreetnanda@gmail.com>
+ * @copyright Jalpreet Singh Nanda (:imjalpreet) 2016
+ */
+
+namespace OCA\MediaMetadata\Hooks;
+
+function getimagesize($filename = null, array &$imageinfo = null) {
+	return ImageHooksTest::$dimensions ?: getimagesize($filename);
```
Comment (oparoz) [inline tests/unit/Hooks/ImageHooksTest.php]: You don't need the check here.

1) we want it to always return the static dimensions
2) the filename will never exist

File: tests/unit/Hooks/ImageHooksTest.php (hunk 12 above)
Comment (imjalpreet) [inline tests/unit/Hooks/ImageHooksTest.php]: Okay. :+1:

File: tests/unit/Hooks/ImageHooksTest.php (hunk 11 above)
Comment (oparoz) [inline tests/unit/Hooks/ImageHooksTest.php]: Something must have changed in your PHP env. Try maybe to use PHP 7?
If it seg faults here, chances are it will seg fault as well when using the app

File: tests/unit/Hooks/ImageHooksTest.php (hunk 13 above)
Comment (imjalpreet) [inline tests/unit/Hooks/ImageHooksTest.php]: Ah.. This was the reason for the segfault. But now I am getting this expection failed error: [link](https://gist.github.com/imjalpreet/54e6ce7ddff35455a370ab9238149830)
Is this because the way `entity` is mocked is wrong?

File: tests/unit/Hooks/ImageHooksTest.php (hunk 11 above)
Comment (oparoz) [inline tests/unit/Hooks/ImageHooksTest.php]: Ah, one thing you could try is to disable the opcache.

File: tests/unit/Hooks/ImageHooksTest.php (hunk 13 above)
Comment (oparoz) [inline tests/unit/Hooks/ImageHooksTest.php]: That's much better and yes, that's because of the mapper. You can see that you have 2 completely different objects: a mocked one and a real one.

**Comment by imjalpreet** [comment]
@oparoz I have taken reference from core and updated the test but it is still not working and giving the following failure message: `Expectation failed for method name is equal to <string:lastInsertId> when invoked 1 time(s).
Method was expected to be called 1 times, actually called 0 times.
`
Also, `[PHPUnit_Framework_Exception] Argument #2 (No Value) of PHPUnit_Framework_Assert::assertContains() must be a array, traversable or string` at line 110 ImageHooksTest

Can you help me out?

Reference:
[Link 1](https://github.com/owncloud/core/blob/master/tests/lib/AppFramework/Db/EntityTest.php)
[Link 2](https://github.com/owncloud/core/blob/master/tests/lib/AppFramework/Db/MapperTest.php)

**Comment by imjalpreet** [comment]
The assertions start to pass if I use the real ImageDimension object but the expects lastInsertId does not pass even then.

**Comment by oparoz** [comment]
> I have taken reference from core and updated the test but it is still not working

The problem is that you seem to have just copied/pasted the method containing `lastInsertId` without really understanding what it does. You need to make modifications so that it applies to your method.

File: tests/unit/Hooks/ImageHooksTest.php [hunk 14]
```diff
@@ -99,19 +102,18 @@ public function testPostCreate() {
...
-			->method('insert')
-			->with($this->imageDimension);
+		$this->db->expects($this->once())
+			->method('lastInsertId')
+			->with($this->equalTo('*PREFIX*table'));
+
+		$this->imageDimension->setId($jpgFile->getId());
+		$this->assertContains('id', $this->imageDimension->getUpdatedFields());
+		$this->imageDimension->setImageHeight($imageHeight);
+		$this->assertEquals(2, count($this->imageDimension->getUpdatedFields()));
```
Comment (oparoz) [inline tests/unit/Hooks/ImageHooksTest.php]: This is too vague, you need to check the values of the row. In the example you gave me, it shows that a new fake TestEntity class was created in order to test the insertion of rows. You should try to do the same.

Comment (imjalpreet) [inline tests/unit/Hooks/ImageHooksTest.php]: @oparoz Actually, right now I had written this only to check if it was working but I had planned to test the values once things start working. 
Also, regarding the TestEntity class, I thought I can test it with the ImageDimension class as it is also an extension of the entity class. Do I have to create a fake TestEntity class? In that case, how will we test the ImageDimension class?

**Comment by imjalpreet** [comment]
@oparoz,

> The problem is that you seem to have just copied/pasted the method containing lastInsertId without really understanding what it does. You need to make modifications so that it applies to your method.

Actually, I had used it after understanding it and I thought the way I am using it is correct. My understanding was that we will be able to successfully test the insert method if we check whether lastInsertId is called once as it the method which is called in the insert method. Did I understand it wrong?

File: lib/Services/ImageDimensionMapper.php [hunk 15]
```diff
@@ -14,12 +14,13 @@
...
+use OCP\IDBConnection;
 
 class ImageDimensionMapper extends Mapper {
 	/**
-	 * @param IDb $database
+	 * @param IDBConnection $database
 	 */
-	public function __construct(IDb $database) {
+	public function __construct(IDBConnection $database) {
 		parent::__construct($database, 'mediametadata_image_size', '\OCA\MediaMetadata\Services\ImageDimension');
```
Comment (oparoz) [inline lib/Services/ImageDimensionMapper.php]: Please use Dependency Injection here. There is an explanation as to why in the official documentation.

File: tests/unit/Hooks/ImageHooksTest.php (hunk 14 above)
Comment (oparoz) [inline tests/unit/Hooks/ImageHooksTest.php]: > Actually, right now I had written this only to check if it was working but I had planned to test the values once things start working. 

OK, all good then.

> Also, regarding the TestEntity class, I thought I can test it with the ImageDimension class as it is also an extension of the entity class.

It is in the real class, but in your test it isn't. Methods like `getUpdatedFields` don't exist in your mock class and you won't get any results back.
If I'm right, you should build your ImageDimension mock like TestEntity in the example.

**Comment by oparoz** [comment]
> Actually, I had used it after understanding it and I thought the way I am using it is correct. My understanding was that we will be able to successfully test the insert method if we check whether lastInsertId is called once as it the method which is called in the insert method. Did I understand it wrong?

You may be right and the reason it isn't called is because your `ImageDimension` mock is not working. Let's see what happens when you fix that problem.

**Comment by imjalpreet** [comment]
@oparoz I will do the required changes and get back to you.

**Comment by imjalpreet** [comment]
@oparoz I have done something similar to testEntity, it is now passing the assertions but lastInsertId is still not being called. I have tried a few things but it is still failing.

**Comment by imjalpreet** [comment]
@oparoz I had even tried to do something similar to exampleMapper in MapperTest.php but it was giving the following failure message: `Running OCA\MediaMetadata\Hooks\ImageHooksTest::testPostCreat... PHP Fatal error:  Call to a member function bindValue() on a non-object in /var/www/GSoC/lib/public/AppFramework/Db/Mapper.php on line 246`

**Comment by oparoz** [comment]
This error tells me that something is not mocked properly. You probably need to add that `bindValue` method to the mocked object.

**Comment by imjalpreet** [comment]
@oparoz I will try adding that method and see if its working.  Also, I have added the database field names for gps fields.

**Comment by imjalpreet** [comment]
@oparoz Can you help me on how to add that method to the mocked object? I have tried but I am not successful and also I am not sure whether the error is due to that. 

I would also like to inform you that I have also started with extraction of priority one fields for exif data.

**Comment by oparoz** [comment]
I will help you asap. I've been unable to look at this these past days.

In the mean time, we need to split this PR in two because it's holding things back.

Please submit another PR containing everything except the `ImageHooks->postCreate` test and once I've merged it, you can then rebase this one.

**Comment by imjalpreet** [comment]
@oparoz Okay, I will do that.

**Comment by imjalpreet** [comment]
@oparoz Should I remove the entire ImageHooksTest file?

**Comment by oparoz** [comment]
> Should I remove the entire ImageHooksTest file?

Makes sense too

**Comment by oparoz** [comment]
Next steps:
- [x] Merge PR with test setup
- [x] Rebase
- [x] Push changes which include exampleMapper

**Comment by imjalpreet** [comment]
@oparoz I have added the ImageHooksTest with exampleMapper and as I had told you earlier it is still giving the same bindValue error.

**Comment by oparoz** [comment]
OK, I've found the problem: You're missing an entire class: https://github.com/owncloud/core/blob/master/tests/lib/AppFramework/Db/MapperTest.php#L55

But I think we need to design this differently

The hook should send the file to an Extractor service which then sends the extracted data to a DB inserting service. That way we keep all these DB testing methods for the DB service.

The Extractor service should be instantiated with the DB service and various extractors classes (Exif, etc.).

Extractor->extract would call various methods to extract and convert the data we need. The next step would be to send the collected data to the DB service and finally it would return an array or a boolean, sending back the data or just `false`.

The DB service would simply insert the data and return a boolean.

What do you think so far?

**Comment by imjalpreet** [comment]
@oparoz I think that's a pretty good idea. So, should I start to implement this method? or do you want to tell something more which may be important and related to this? Otherwise, I will start with this.

**Comment by oparoz** [comment]
Great, then you can start right away, the idea is to have classes which focus on single tasks and methods which are easier to test.
I'll be on IRC today if you have any questions.

**Comment by imjalpreet** [comment]
@oparoz Sure, I will come to IRC if I have any doubts.

**Comment by oparoz** [comment]
Let's close this one.
